### PR TITLE
Assembler Level-Up: Use new patterns under feature flag

### PR DESCRIPTION
**Pull request by miksansegundo**
Related to https://github.com/Automattic/wp-calypso/issues/84893

## Proposed Changes

* Uses the feature flag `pattern-assembler/v2`
* We'll show all and only patterns from the new source site https://assemblerv2patterns.wordpress.com
  * No tags are needed for new patterns until they are moved to Dotcompatterns 
* Force using content from Assembler theme demo site to help test with existing sites

### Important notes 
- Patterns from dotcompatterns.wordpress.com that we want in Assembler v2 must be copied into the new source site
- Page patterns also have to be copied into the new source site and tagged with the tag `Assembler_page`

### Assembler on Calypso Live

https://github.com/Automattic/wp-calypso/assets/1881481/acaeed57-b20c-4b3d-9f8d-9bb4b8e16346

## Testing Instructions

* Sandbox D131496-code
* Access the Assembler using the Calypso Live link below
  * Click "Calypso Live (direct link)" and then create a new site, then continue until the Design picker and click "Design your own" to access the assembler.
  * Another option is using an existing site, click the same link, and then update the URL to be `/setup/with-theme-assembler/patternAssembler?siteSlug=YOUR-SITE-URL`
* Add the param `flags=pattern-assembler/v2` to the URL
* Verify you see only the patterns from assemblerv2patterns.wordpress.com
* Add some patterns
* Verify there are blank patterns because they contain a reference to a `wp_block` pattern
* Verify the styles and fonts work as expected - The variations are still fetched from Creatio 2
* Verify pages are still not available
* Continue until the editor and verify the Home template is created as expected

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by richtabor** [comment]
> Note for designers: feel free to remove this demo pattern

I updated it to a proper header pattern.

File: config/wpcalypso.json
Comment (taipeicoder) [inline config/wpcalypso.json]: [nit] It would be better to align with the feature flag `pattern-assembler/v2` and only https://github.com/Automattic/wp-calypso/pull/84960. WDYT?

**Comment by taipeicoder** [comment]
Proposal to use a different source site for the new patterns instead of using another tag: p1702001999281829-slack-CRWCHQGUB

**Comment by fushar** [comment]
> When the feature flag is enabled, show v1 priority patterns in "Show more patterns"

Apart from the ongoing discussion, do we think the existing patterns will show nicely under the new Assembler theme? I skimmed briefly on Horizon and many of them have "broken" paddings 😢

**Comment by taipeicoder** [comment]
> > When the feature flag is enabled, show v1 priority patterns in "Show more patterns"
> 
> Apart from the ongoing discussion, do we think the existing patterns will show nicely under the new Assembler theme? I skimmed briefly on Horizon and many of them have "broken" paddings 😢

I think a significant number of them doesn't look great, mostly because of the paddings and several simply not rendering 🙁
I wonder that when we launch the Assembler using the theme Assembler, we can just display the new patterns, the previous assembler priority patterns (which needs updating), and don't fetch the rest for now.

File: config/wpcalypso.json
Comment (miksansegundo) [reply config/wpcalypso.json]: I will revert this.

It's here to help testing without passing the flag in the URL.

**Comment by richtabor** [comment]
> I wonder that when we launch the Assembler using the theme Assembler, we can just display the new patterns, the previous assembler priority patterns (which needs updating), and don't fetch the rest for now.

That's probably what we'll need to do. We can port over the patterns we need by adding them to the v2 source site and making the necessary edits.

**Comment by miksansegundo** [comment]
I pushed a new version that uses only patterns from assemblerv2patterns.wordpress.com.

I had to change the PTK and the block renderer APIs to support the new source site. After doing that, I realized that some patterns were blank on the Assembler because the new source site has synced patterns (`wp_block` post type) with a `ref` specific to that site.

After messaging @richtabor to understand how they are working with patterns, basically using the core way, I decided to give it a try to render `wp_block` patterns in the Assembler, and with some more changes on the PTK API they work fine. 

This is exciting news and a game changer for the future of Dotcom patterns and the Assembler. 

`wp_block` patterns cannot use tags, only categories. To identify the page patterns that we show on the pages screen in Assembler we can use the category name "Pages" as TT4 does. 

I have reported to Rich some fixable issues in some of the patterns in assemblerv2patterns.wordpress.com:

- a missing color on a header button
- a pattern that contains a query block - The other pattern with a query block works so probably there is something wrong in that query
- some images and a video that point to pattern-assembler.local
- some gallery patterns have placeholder images that are empty and use style="aspect-ratio:4/3;object-fit:cover"
- we can improve the pattern names

Here is a video demo:

https://github.com/Automattic/wp-calypso/assets/1881481/65bb96b0-b903-4c92-89ab-935e47eb4021

The server-side changes are in the diff D131496-code 

I've updated the description too.

File: config/horizon.json
Comment (miksansegundo) [inline config/horizon.json]: I will revert this.

See https://github.com/Automattic/wp-calypso/pull/85025

Comment (taipeicoder) [reply config/horizon.json]: @miksansegundo do you mean you'll revert this after it's been approved and before merging to trunk?

Comment (fushar) [reply config/horizon.json]: Should be, because the CfT for this feature is already running :)

File: packages/data-stores/src/site/constants.ts
Comment (arthur791004) [inline packages/data-stores/src/site/constants.ts]: This may not work if the file is loaded before the calypso-config is initialized

**Comment by autumnfjeld** [comment]
https://github.com/Automattic/wp-calypso/pull/84977#issuecomment-1846991594
Agree with @richtabor  & @taipeicoder .

File: packages/data-stores/src/site/constants.ts
Comment (miksansegundo) [reply packages/data-stores/src/site/constants.ts]: I'm understanding that you know that case will happen and we have to find another way to use a different constant. Is there another way to do that?

Comment (fushar) [reply packages/data-stores/src/site/constants.ts]: Another way is to turn the constant into function, like: https://github.com/Automattic/wp-calypso/pull/84960

File: config/horizon.json
Comment (miksansegundo) [reply config/horizon.json]: I realized that when I was AFK. I'll revert that today.

File: packages/data-stores/src/site/constants.ts
Comment (miksansegundo) [reply packages/data-stores/src/site/constants.ts]: Thanks 😊

**Comment by fushar** [comment]
Can we update the description to reflect the changes (re: post type and feature flags)?

**Comment by miksansegundo** [comment]
I have updated this PR only to support `post` type patterns. It's expected to see blank patterns when the post contains a `ref` to a `wp_block` pattern.


Screenshot 2566-12-12 at 12 59 40

Designers can still keep working with `wp_block` patterns in assemblerv2patterns.wordpress.com and decide when to export them as `post` patterns by manually copying the markup of each `wp_block` pattern into their corresponding `post` pattern. cc: @autumnfjeld @alaczek @richtabor

**Comment by miksansegundo** [comment]
> Can we update the description to reflect the changes (re: post type and feature flags)?

@fushar, I just updated it. Let me know if there is anything missing.

**Review by fushar — DISMISSED** [review]
Works perfectly 🙏

File: packages/data-stores/src/site/constants.ts
Comment (miksansegundo) [inline packages/data-stores/src/site/constants.ts]: How can we avoid having `getPatternSourceSiteID` here and also in "pattern-assembler/constants.ts"?

Comment (miksansegundo) [reply packages/data-stores/src/site/constants.ts]: Oh, I see now that they are different functions.

Comment (fushar) [reply packages/data-stores/src/site/constants.ts]: Oops, I missed this haha. Yeah, we need to create two functions!

File: packages/data-stores/src/site/constants.ts
Comment (fushar) [inline packages/data-stores/src/site/constants.ts]: We can now remove this right?

File: packages/data-stores/src/site/constants.ts
Comment (miksansegundo) [reply packages/data-stores/src/site/constants.ts]: I turned the constant to a function named `getPlaceholderSiteID` and checked that it's only been used in the assembler. See https://github.com/search?q=repo%3AAutomattic%2Fwp-calypso%20PLACEHOLDER_SITE_ID&type=code

**Comment by fushar** [comment]
Hmmm, I just realized that this might interfere with @arthur791004's CfT on Horizon for the Assembler-First flow... it uses the old Creatio 2 theme but with this PR we fetch the patterns from the new source site + using the new placeholder site, which will look weird.

Let's think a bit more...

File: packages/data-stores/src/site/constants.ts
Comment (miksansegundo) [reply packages/data-stores/src/site/constants.ts]: Done

**Comment by taipeicoder** [comment]
> Hmmm, I just realized that this might interfere with @arthur791004's CfT on Horizon for the Assembler-First flow... it uses the old Creatio 2 theme but with this PR we fetch the patterns from the new source site + using the new placeholder site, which will look weird.
> 
> Let's think a bit more...

Given that the `assembler-first` flow is also an internal CfT (ends this Friday), and the feature doesn't overlap with this one, could we just:

- Update the `assembler-first` flow to use the Assembler theme.
- Update the `assembler-first` CfT to let testers know to "ignore" patterns and styles.

What do you all think? @autumnfjeld @arthur791004 @Automattic/lego

**Comment by fushar** [comment]
Agree for the simplicity 👍 

> Update the assembler-first flow to use the Assembler theme.

Let's also do this in this PR I think (by "reverting" https://github.com/Automattic/wp-calypso/pull/85025).

**Comment by miksansegundo** [comment]
> Hmmm, I just realized that this might interfere with @arthur791004's CfT on Horizon for the Assembler-First flow
Only when testing with existing sites.

That CfT ends on Dec 15, 2023; Maybe we just wait to merge this PR once the CfT is completed?

**Comment by fushar** [comment]
> That CfT ends on Dec 15, 2023; Maybe we just wait to merge this PR once the CfT is completed?

This project's ETA to put everything in Horizon is on Dec 13 🥲 pbxlJb-56T-p2

**Comment by miksansegundo** [comment]
> Given that the assembler-first flow is also an internal CfT (ends this Friday), and the feature doesn't overlap with this one, could we just:
> Update the assembler-first flow to use the Assembler theme.
> Update the assembler-first CfT to let testers know to "ignore" patterns and styles.

Sounds good to me 👍

**Comment by miksansegundo** [comment]
> Let's also do this in this PR I think (by "reverting" https://github.com/Automattic/wp-calypso/pull/85025).

@fushar, what should be reverted? I think that function is still required, and the condition will work as we want after we enable the flag on Horizon.

```
if ( isEnabled( 'pattern-assembler/v2' ) ) {
  return ASSEMBLER_V2_DESIGN;
}
```

**Comment by fushar** [comment]
> @fushar, what should be reverted? I think that function is still required, and the condition will work as we want after we enable the flag on Horizon.

@miksansegundo This change needs to be reverted so that the assembler-first flow uses the Assembler theme:

image

**Review by fushar — APPROVED** [review]

**Comment by miksansegundo** [comment]
>Given that the assembler-first flow is also an internal CfT (ends this Friday), and the feature doesn't overlap with this one, could we just:
>
>Update the assembler-first flow to use the Assembler theme.
>Update the assembler-first CfT to let testers know to "ignore" patterns and styles.

I updated the assembler-first flow to use the Assembler theme in https://github.com/Automattic/wp-calypso/pull/84977/commits/39e48293e4db60febdb9015fb1761fb6dd67cdae and updated the CfT to let testers know to "ignore" theme, patterns and styles.

When testing the assembler-first flow, I noticed that the theme is not applied if using a non-automattician account because it's only soft-launched. Also, most of the v2 patterns are still blank because they should not include references to `wp_block` patterns. These issues are expected for now. 

cc: @autumnfjeld @arthur791004 @Automattic/lego

**Comment by miksansegundo** [comment]
It seems that we cannot enable the Assembler theme on Horizon for now because we are still testing another feature (Assembler-first flow) that needs to be tested with Creatio 2 because it's a theme launched for all users.

I have reverted this commit https://github.com/Automattic/wp-calypso/pull/84977/commits/39e48293e4db60febdb9015fb1761fb6dd67cdae, and now the Assembler-first CfT can continue using Creatio 2.

See pekYwv-3vI-p2#comment-2715

**Comment by taipeicoder** [comment]
I feel that this is more complicated than it needs to be 😕
@Automattic/lego are there any concerns to merge this after the assembler-first flow CfT is closed (aka next week) as @miksansegundo proposed in https://github.com/Automattic/wp-calypso/pull/84977#issuecomment-1851423631? Seems to me that it's more trouble than needed trying to get this on Horizon without affecting the assembler-first flow CfT. Just my two cents cc: @autumnfjeld 

EDIT: Another approach would be to put it on Horizon, but not enable the flag `pattern-assembler/v2` by default. So essentially: `https://horizon.wordpress.com/setup/site-setup/patternAssembler?siteSlug=${site_slug}&flags=pattern-assembler/v2,-themes/assembler-first`

**Comment by autumnfjeld** [comment]
> Another approach would be to put it on Horizon, but not enable the flag pattern-assembler/v2 by default. So essentially: https://horizon.wordpress.com/setup/site-setup/patternAssembler?siteSlug=${site_slug}&flags=pattern-assembler/v2,-themes/assembler-first

Yes, please do this if possible @miksansegundo  @fushar . 😄  Thank you.

There is urgency in getting this on horizon so @alaczek  can do some vetting this week and flag any problems , particularly verifying that we've made the right call here pbxlJb-5ax-p2#comment-3385

**Comment by autumnfjeld** [comment]
Or perhaps enable `assembler/v2` on  wpcalypso.wordpress.com/ ?

**Pull request merged**
